File: Binary Search/Medium/LET_no1011_Capacity To Ship Packages Within D Days.py
class Solution:
    def shipWithinDays(self, weights: List[int], days: int) -> int:
        left, right = max(weights), sum(weights)
        while left + 1 < right:
            mid = (left + right) // 2
            if self.count(weights, mid) > days:
                left = mid
            else:
                right = mid
                
        if self.count(weights, left) <= days:
            return left
        return right
    
    def count(self, weights, limit):
        res = cur = 0
        for w in weights:
            if cur + w > limit:
                res += 1
                cur = 0
            cur += w
        return res + 1

File: Binary Search/Medium/LET_no1011_Capacity To Ship Packages Within D Days.cpp
class Solution {
public:
    int shipWithinDays(vector<int>& weights, int days) {
        int min_cap = *max_element(weights.begin(), weights.end());
        int max_cap = accumulate(weights.begin(), weights.end(), 0);

        while (min_cap + 1 < max_cap) {
            int cap = min_cap + (max_cap - min_cap) / 2;
            cout << cap << endl;
            int count_days = counter(weights, cap);
            if (count_days > days)
                min_cap = cap;
            else
                max_cap = cap;
        }

        if (counter(weights, min_cap) <= days)
            return min_cap;
        return max_cap;
    }

    int counter(vector<int>& weights, int limit) {
        int res = 0, cur_cap = 0;
        for (int w : weights) {
            if (cur_cap + w > limit) {
                res++;
                cur_cap = 0;
            }
            cur_cap += w;
        }

        return res + 1;
    }
};